File: app/src/main/java/knf/kuma/commons/CastUtil.java
package knf.kuma.commons;

import android.app.Activity;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.Looper;

import com.afollestad.materialdialogs.MaterialDialog;

import es.munix.multidisplaycast.CastControlsActivity;
import es.munix.multidisplaycast.CastManager;
import es.munix.multidisplaycast.interfaces.CastListener;
import es.munix.multidisplaycast.interfaces.PlayStatusListener;
import xdroid.toaster.Toaster;

/**
 * Created by Jordy on 25/01/2018.
 */

public class CastUtil implements CastListener, PlayStatusListener {
    public static String NO_PLAYING = "no_play";
    private static CastUtil ourInstance;
    private Context context;
    private MutableLiveData<String> castingEid = new MutableLiveData<>();
    private boolean isConnected = false;

    private MaterialDialog loading;

    private CastUtil(Context context) {
        this.context=context;
        CastManager.getInstance().setDiscoveryManager();
        CastManager.getInstance().setPlayStatusListener(getClass().getSimpleName(),this);
        CastManager.getInstance().setCastListener(getClass().getSimpleName(),this);
        castingEid.setValue(NO_PLAYING);
    }

    public static void init(Context context) {
        ourInstance = new CastUtil(context);
    }

    public static CastUtil get() {
        return ourInstance;
    }

    public boolean connected() {
        return isConnected;
    }

    public MutableLiveData<String> getCasting(){
        return castingEid;
    }

    public void play(Activity activity,String eid, String url, String title, String chapter, String preview, boolean isAid){
        if (connected()){
            startLoading(activity);
            setEid(eid);
            if (isAid)
                preview="https://animeflv.net/uploads/animes/thumbs/"+preview+".jpg";
            CastManager.getInstance().playMedia(url,"video/mp4",title,chapter,preview);
        }else {
            Toaster.toast("No hay dispositivo seleccionado");
        }
    }

    public void onDestroy() {
        CastManager.getInstance().unsetCastListener(getClass().getSimpleName());
        CastManager.getInstance().unsetPlayStatusListener(getClass().getSimpleName());
        CastManager.getInstance().onDestroy();
    }

    @Override
    public void isConnected() {
        isConnected = true;
    }

    @Override
    public void isDisconnected() {
        isConnected = false;
    }

    private MaterialDialog getLoading(Activity activity) {
        loading = new MaterialDialog.Builder(activity)
                .content("Cargando...")
                .progress(true, 0)
                .cancelable(false)
                .build();
        return loading;
    }

    private void startLoading(final Activity activity){
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                getLoading(activity).show();
            }
        });
    }

    private void stopLoading(){
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                if (loading!=null)
                    loading.dismiss();
            }
        });
    }

    private void setEid(final String eid){
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                castingEid.setValue(eid);
            }
        });
    }

    public void openControls(){
        context.startActivity(new Intent(context, CastControlsActivity.class));
    }

    @Override
    public void onPlayStatusChanged(int playStatus) {
        switch (playStatus) {
            case STATUS_START_PLAYING:
                stopLoading();
                openControls();
                break;
            case STATUS_FINISHED:
            case STATUS_STOPPED:
                setEid(NO_PLAYING);
                break;
            case STATUS_NOT_SUPPORT_LISTENER:
                stopLoading();
                setEid(NO_PLAYING);
                Toaster.toast("Video no soportado por diapositivo");
                break;
        }
    }

    @Override
    public void onPositionChanged(long currentPosition) {

    }

    @Override
    public void onTotalDurationObtained(long totalDuration) {

    }

    @Override
    public void onSuccessSeek() {

    }
}

File: app/src/main/java/knf/kuma/commons/CastUtil.kt
package knf.kuma.commons

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.util.Log
import android.view.Menu
import android.view.View
import androidx.lifecycle.MutableLiveData
import com.google.android.material.snackbar.Snackbar
import es.munix.multidisplaycast.CastManager
import es.munix.multidisplaycast.interfaces.CastListener
import es.munix.multidisplaycast.interfaces.PlayStatusListener
import knf.kuma.App
import knf.kuma.achievements.AchievementManager
import knf.kuma.cast.CastCustom
import knf.kuma.cast.CastMedia
import knf.kuma.custom.ThemedControlsActivity
import org.jetbrains.annotations.Contract
import xdroid.toaster.Toaster

class CastUtil private constructor(private val context: Context) : CastListener, PlayStatusListener {
    val casting = MutableLiveData<String>()

    private var loading: Snackbar? = null

    init {
        CastManager.setInstance(CastCustom())
        CastManager.getInstance().setDiscoveryManager()
        CastManager.getInstance().setPlayStatusListener(javaClass.simpleName, this)
        CastManager.getInstance().setCastListener(javaClass.simpleName, this)
        casting.value = NO_PLAYING
    }

    fun registerActivity(activity: Activity, menu: Menu, menuId: Int) = CastManager.getInstance().registerForActivity(activity, menu, menuId)

    fun connected(): Boolean {
        return CastManager.getInstance().isConnected
        //return isConnected;
    }

    fun play(view: View, castMedia: CastMedia?) {
        try {
            if (castMedia == null) throw IllegalStateException("CastMedia must not be null")
            if (connected()) {
                if (!castMedia.url.endsWith(":" + SelfServer.HTTP_PORT))
                    SelfServer.stop()
                startLoading(view)
                setEid(castMedia.eid)
                Log.e("Cast", castMedia.url)
                CastManager.getInstance().playMedia(castMedia.url, "video/mp4", castMedia.title, castMedia.subTitle, castMedia.image)
                AchievementManager.unlock(listOf(6))
            } else {
                Toaster.toast("No hay dispositivo seleccionado")
            }
        } catch (e: Exception) {
            stopLoading()
            Toaster.toast("Error al reproducir")
        }
    }

    fun stop() {
        CastManager.getInstance().stop()
    }

    fun onDestroy() {
        loading?.safeDismiss()
        loading = null
        CastManager.getInstance().onDestroy()
    }

    override fun isConnected() {
    }

    override fun isDisconnected() {
        stopLoading()
        setEid(NO_PLAYING)
        SelfServer.stop()
    }

    private fun getLoading(view: View): Snackbar {
        return view.showSnackbar("Cargando...", duration = Snackbar.LENGTH_INDEFINITE)
    }

    private fun startLoading(view: View) {
        doOnUI {
            loading = getLoading(view)
        }
    }

    private fun stopLoading() {
        doOnUI {
            loading?.safeDismiss()
            loading = null
        }
    }

    private fun setEid(eid: String) {
        doOnUI {
            casting.setValue(eid)
        }
    }

    fun openControls() {
        context.startActivity(Intent(context, ThemedControlsActivity::class.java).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
    }

    override fun onPlayStatusChanged(playStatus: Int) {
        when (playStatus) {
            PlayStatusListener.STATUS_START_PLAYING -> {
                stopLoading()
                openControls()
            }
            PlayStatusListener.STATUS_FINISHED, PlayStatusListener.STATUS_STOPPED -> {
                stopLoading()
                setEid(NO_PLAYING)
            }
            PlayStatusListener.STATUS_NOT_SUPPORT_LISTENER -> {
                stopLoading()
                setEid(NO_PLAYING)
                Toaster.toast("Video no soportado por dispositivo")
            }
        }
    }

    override fun onPositionChanged(currentPosition: Long) {

    }

    override fun onTotalDurationObtained(totalDuration: Long) {

    }

    override fun onSuccessSeek() {

    }

    companion object {
        var NO_PLAYING = "no_play"
        private val ourInstance: CastUtil by lazy { CastUtil(App.context) }

        @Contract(pure = true)
        fun get(): CastUtil {
            return ourInstance
        }

        fun registerActivity(activity: Activity, menu: Menu, menuId: Int) = get().registerActivity(activity, menu, menuId)
    }
}
